config/archive.go

```diff
@@ -0,0 +1,86 @@
+package config
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/contiv/volplugin/errors"
+	"github.com/contiv/volplugin/watch"
+	"github.com/coreos/etcd/client"
+	"golang.org/x/net/context"
+)
+
+func (c *Client) policyArchive(name string) string {
+	return c.prefixed(rootPolicyArchive, name)
+}
+
+func (c *Client) policyArchiveEntry(name string, timestamp string) string {
+	return c.prefixed(rootPolicyArchive, name, timestamp)
+}
+
+// CreatePolicyRevision creates an revision entry in a policy's history.
+func (c *Client) CreatePolicyRevision(name string, policy string) error {
+	timestamp := fmt.Sprint(time.Now().Unix())
+	key := c.policyArchiveEntry(name, timestamp)
+
+	_, err := c.etcdClient.Set(context.Background(), key, policy, &client.SetOptions{PrevExist: client.PrevIgnore})
+	if err != nil {
+		return errors.EtcdToErrored(err)
+	}
+
+	return nil
+}
+
+// ListPolicyRevisions returns a list of all the revisions for a given policy.
+func (c *Client) ListPolicyRevisions(name string) ([]string, error) {
+	keyspace := c.policyArchive(name)
+
+	resp, err := c.etcdClient.Get(context.Background(),
+		keyspace,
+		&client.GetOptions{Sort: true, Recursive: false, Quorum: true},
+	)
+	if err != nil {
+		return nil, errors.EtcdToErrored(err)
+	}
+
+	ret := []string{}
+
+	for _, node := range resp.Node.Nodes {
+		parts := strings.Split(node.Key, "/")
+
+		ret = append(ret, parts[len(parts)-1])
+	}
+
+	return ret, nil
+}
+
+// GetPolicyRevision returns a single revision for a given policy.
+func (c *Client) GetPolicyRevision(name, revision string) (string, error) {
+	keyspace := c.policyArchiveEntry(name, revision)
+
+	resp, err := c.etcdClient.Get(context.Background(),
+		keyspace,
+		&client.GetOptions{Sort: false, Recursive: false, Quorum: true},
+	)
+	if err != nil {
+		return "", errors.EtcdToErrored(err)
+	}
+
+	return resp.Node.Value, nil
+}
+
+// WatchForPolicyChanges creates a watch which prints the policy name and revision number
+// (timestamp) to stdout whenever a policy is uploaded.
+func (c *Client) WatchForPolicyChanges(watchChan chan *watch.Watch) {
+	w := watch.NewWatcher(watchChan, c.prefixed(rootPolicyArchive), func(resp *client.Response, w *watch.Watcher) {
+		parts := strings.Split(resp.Node.Key, "/")
+
+		policy := parts[0]
+		timestamp := parts[1]
+
+		fmt.Printf("%s\t%s\n", policy, timestamp)
```

Comment:
I guess this function should not really be directly outputting to stdout but rather returning just returning the info about *what* was changed and letting the creator of the watch decide what to do with it.